### PR TITLE
lockdep: fix follows/follows_bt resize() size

### PR DESCRIPTION
**Pull request by lxbsz**
When lock number > 4096, it will cause crash like:

```
980038  ceph version 17.0.0-1852-g7d8ba51fbd0 (7d8ba51fbd0258713432a7a0950d258fb9d833b2) quincy (dev)
980039  1: ./bin/ceph-fuse(+0x64c8a6) [0x563b97a4c8a6]
980040  2: /lib64/libpthread.so.0(+0x12dc0) [0x147c2574bdc0]
980041  3: gsignal()
980042  4: abort()
980043  5: (operator new(unsigned long, void*)+0) [0x147c2784b790]
980044  6: (std::vector<std::bitset<4096ul>, std::allocator<std::bitset<4096ul> > >::operator[](unsigned long)+0x4f) [0x147c27ab0ee7       ]
980045  7: (lockdep_will_lock(char const*, int, bool, bool)+0x58f) [0x147c27aae843]
980046  8: (ceph::mutex_debug_detail::mutex_debugging_base::_will_lock(bool)+0x4a) [0x147c27abdb48]
980047  9: (ceph::mutex_debug_detail::mutex_debug_impl<false>::lock(bool)+0x38) [0x563b978962e0]
980048  10: (std::scoped_lock<ceph::mutex_debug_detail::mutex_debug_impl<false> >::scoped_lock(ceph::mutex_debug_detail::mutex_debug_impl<       false>&)+0x2f) [0x563b9789440f]
980049  11: (Client::add_update_cap(Inode*, MetaSession*, unsigned long, unsigned int, unsigned int, unsigned int, unsigned int, inodeno_t       , int, UserPerm const&)+0xd0) [0x563b978e35ba]
980050  12: (Client::add_update_inode(InodeStat*, utime_t, MetaSession*, UserPerm const&)+0x11c2) [0x563b978c0fb0]
980051  13: (Client::insert_trace(MetaRequest*, MetaSession*)+0x1014) [0x563b978c55c4]
980052  14: (Client::handle_client_reply(boost::intrusive_ptr<MClientReply const> const&)+0xb3c) [0x563b978d22c2]
980053  15: (Client::ms_dispatch2(boost::intrusive_ptr<Message> const&)+0x4e0) [0x563b978d3f8c]
980054  16: (Messenger::ms_deliver_dispatch(boost::intrusive_ptr<Message> const&)+0xe9) [0x147c27bb1add]
980055  17: (DispatchQueue::entry()+0x61d) [0x147c27bb04c9]
980056  18: (DispatchQueue::DispatchThread::entry()+0x1c) [0x147c27d35ad6]
980057  19: (Thread::entry_wrapper()+0x83) [0x147c279675a7]
980058  20: (Thread::_entry_func(void*)+0x18) [0x147c2796751a]
980059  21: /lib64/libpthread.so.0(+0x82de) [0x147c257412de]
980060  22: clone()
```

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/common/lockdep.cc [hunk 1]
```diff
@@ -41,9 +41,8 @@ static constexpr size_t MAX_LOCKS = 128 * 1024;   // increase me as needed
 static std::bitset<MAX_LOCKS> free_ids; // bit set = free
 static ceph::unordered_map<pthread_t, std::map<int,ceph::BackTrace*> > held;
 static constexpr size_t NR_LOCKS = 4096; // the initial number of locks
-static constexpr size_t MAX_FOLLOWERS = 4096;
-static std::vector<std::bitset<MAX_FOLLOWERS>> follows(NR_LOCKS); // follows[a][b] means b taken after a
-static std::vector<std::array<ceph::BackTrace *, MAX_FOLLOWERS>> follows_bt(NR_LOCKS);
+static std::vector<std::bitset<MAX_LOCKS>> follows(NR_LOCKS); // follows[a][b] means b taken after a
```
Comment (tchaikov) [inline src/common/lockdep.cc:44]: i don't understand why s/MAX_FOLLOWERS/MAX_LOCKS/. could you expand a little bit on it?

File: src/common/lockdep.cc [hunk 2]
```diff
@@ -41,9 +41,8 @@ static constexpr size_t MAX_LOCKS = 128 * 1024;   // increase me as needed
...
+static std::vector<std::map<int,ceph::BackTrace *>> follows_bt(NR_LOCKS);
```
Comment (tchaikov) [inline src/common/lockdep.cc:45]: yeah! makes sense to me to use `map<>` to store a sparse mapping.

File: src/common/lockdep.cc [hunk 3]
```diff
@@ -173,8 +172,8 @@ static int _lockdep_register(const char *name)
     if (current_maxid <= (unsigned)id) {
       current_maxid = (unsigned)id + 1;
       if (current_maxid == follows.size()) {
-        follows.resize(current_maxid);
-        follows_bt.resize(current_maxid);
+        follows.resize(current_maxid + 1);
+        follows_bt.resize(current_maxid + 1);
```
Comment (tchaikov) [inline src/common/lockdep.cc:176]: yet another off-by-one! sigh..

File: src/common/lockdep.cc (hunk 1 above)
Comment (tchaikov) [reply src/common/lockdep.cc:44]: ahh, i see. the followers can still be distributed to the space higher than 4096, even the total number of which should be much smaller than 4096.

**Review by tchaikov — APPROVED** [review]

File: src/common/lockdep.cc (hunk 1 above)
Comment (lxbsz) [reply src/common/lockdep.cc:44]: `follows[p->first].set(id); `

The maximum of the `id` could reach to `MAX_LOCKS - 1`.

**Review by batrick — APPROVED** [review]

**Comment by tchaikov** [comment]
jenkins test make check

**Review by xxhdx1985126 — APPROVED** [review]

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-03-14_09:40:23-rados-wip-kefu-testing-2021-03-14-1612-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-03-15_02:29:57-rados-wip-kefu-testing-2021-03-15-0034-distro-basic-smithi/

**Comment by tchaikov** [comment]
@batrick the failures are not related. will leave this PR to you.

**Comment by batrick** [comment]
https://pulpito.ceph.com/pdonnell-2021-03-15_22:16:56-fs-wip-pdonnell-testing-20210315.182203-distro-basic-smithi/

- https://tracker.ceph.com/issues/49842
- https://tracker.ceph.com/issues/49466
- https://tracker.ceph.com/issues/49822
- https://tracker.ceph.com/issues/49240
- https://tracker.ceph.com/issues/48773
- https://tracker.ceph.com/issues/45434
- https://tracker.ceph.com/issues/49500
- https://tracker.ceph.com/issues/49843
- https://tracker.ceph.com/issues/49845
- https://tracker.ceph.com/issues/48805
- https://tracker.ceph.com/issues/49605

and failure caused by PR: https://github.com/ceph/ceph/pull/39969

**Pull request merged**